### DISCUSSION TITLE
@urql/svelte - How to reset/invalidate the cache?

### DISCUSSION THREAD

**Discussion by iamlouk** [urql-graphql/urql#2023] [Q&A] [answered]
Hi,

Is there a way to reset the cache or at least some entries from the cache in svelte when using urql? We have a scenario where the values can expire and get outdated, so the same query should rerun and not hit the cache after a certain time. I know that this sounds like something that could need a subscription, but there are other reasons against that. It would be fine to reset the complete cache.

In other threads it is suggested to switch the client, but as in @urql/svelte, `setContext`/`getContext` is used, one can not switch the client without recreating the component. Is there maybe something like an undocumented magic `getClient().cache.invalidateAll()`?

[This](https://github.com/FormidableLabs/urql/discussions/1961) thread asks basically the same question for Vue. 

Thank you very much in advance!

**Comment by kitten** [comment]
This will depend on several factors; Recreating the client and calling `setContext` again will be the most straightforward solution and will reset the entire cache. However, if your goal is to invalidate specific pages _occasionally_ then I suppose this isn't what you're looking for.

Either you're probably looking for our normalised cache, which allows you to update and invalidate data when a mutation is completed, or the more direct method of the [request-policy exchange](https://github.com/FormidableLabs/urql/tree/main/exchanges/request-policy)
This exchange has a more straightforward idea about refetching. It will change any request policy over to `cache-and-network` after a TTL has reached for its next request.

**Reply by iamlouk** [reply]
`setContext` Is not an option as it only works when a component is initialised in Svelte.

I am confused about the `cache-and-network` policy: I understand that at first, the cached result will be returned and the network request will be done in the background. But once newer data has arrived, will the svelte store returned by `operationStore` update again by itself so that the newly fetched data is displayed, or does one have to call `reexecute()` again and trigger the update manually?

**Comment by iamlouk** [comment, marked as answer]
So I decided to write a very simple custom cache exchange, inferior to what already exists in probably most regards, but with the behavior I want. For anyone interested, here it is: https://gist.github.com/iamlouk/ee0d57167d2eb804ec42b0d74cbe36ca (As the application I am working on does quite a lot of very big queries, I also added a `maxSize` to limit memory usage).

I would still like to suggest something like `cacheExchange.clear()` to be added to the existing cache.

**Reply by mlesin** [reply]
Is this exchange able to expire the exchange-graphcache's cache or does it have a completely separate cache?

**Reply by iamlouk** [reply]
It is a completely separate cache, the implementation is very simple, feel free to have a look. The one thing you should know to understand why it works is that `map.values` and `map.keys` return values in insertion order.
